ui/radar/radar.ts

```diff
@@ -0,0 +1,499 @@
+
+      PlaySound(m, options);
+    }
+  }
+
+  UpdateMonsterPuller(monster: Monster, puller: string) {
+    if (!this.options.Puller)
+      return;
+    if (monster.puller || monster.skipPuller)
+      return;
+    monster.puller = puller;
+    this.UpdateMonsterDom(monster);
+    console.log(`Pulled: ${puller} => ${monster.name}`);
+  }
+
+  UpdateMonsterDom(monster: Monster) {
+    // Wait for OnPlayerChange.
+    if (!this.playerPos)
+      return;
+
+    let options = this.options;
+    if (monster.rank && monster.rank in options.RankOptions)
+      options = Object.assign({}, this.options, options.RankOptions[monster.rank]);
+    const tr = monster.dom;
+    // calculate rotation based on facing
+    const playerVector = new Point2D(this.playerPos.x, this.playerPos.y);
+    const targetVector = monster.pos;
+    const deltaVector = new Point2D(
+        targetVector.x - playerVector.x,
+        targetVector.y - playerVector.y,
+    );
+    if (tr) {
+      const node = tr.childNodes[1];
+      if (node && node instanceof HTMLElement) {
+        node.innerHTML = `${monster.rank ?? ''}&nbsp;&nbsp;&nbsp;&nbsp;${monster.name}`;
+        if (Math.abs(this.playerPos.z - monster.posZ) > 5)
+          node.innerHTML += '&nbsp;&nbsp;' + (this.playerPos.z < monster.posZ ? '↑' : '↓');
+        node.innerHTML += '<br>' + deltaVector.length().toFixed(2) + 'm';
+        if (Date.now() / 1000 <= monster.battleTime + 60) {
+          node.innerHTML += ' ' + (monster.currentHp * 100 /
+            monster.hp).toFixed(2) + '%';
+        }
+        if (monster.puller)
+          node.innerHTML += '&nbsp;&nbsp;' + monster.puller;
+        // Z position is relative to the map so it's omitted.
+        if (options.Position) {
+          node.innerHTML += '<br>X: ' +
+            posToMap(monster.pos.x).toFixed(1) + '&nbsp;&nbsp;Y:' +
+            posToMap(monster.pos.y).toFixed(1);
+        }
+      }
+    }
+    if (options.DetectionRange > 0 && deltaVector.length() > options.DetectionRange)
+      monster.dom.classList.add('hide');
+    else
+      monster.dom.classList.remove('hide');
+
+    let deltaTheta = Math.atan2(deltaVector.y, deltaVector.x);
+    deltaTheta -= Math.PI - this.playerRotation;
+    const angle = deltaTheta * 180 / Math.PI;
+    const arrowId = 'arrow-' + monster.id;
+    const arrow = document.getElementById(arrowId);
+    if (arrow)
+      arrow.style.transform = `rotate(${angle}deg)`;
+  }
+
+  RemoveMonster(mobKey: string) {
+    const monster = this.targetMonsters[mobKey];
+    if (!monster)
+      return;
+    console.log(`Killed: ${monster.name}`);
+    monster.dom.remove();
+    delete this.targetMonsters[mobKey];
+  }
+
+  OnNetLog(e: { line: string[]; rawLine: string }) {
+    const type = e.line[0];
+    const log = e.rawLine;
+
+    // added new combatant
+    if (type === '03') {
+      const m = this.regexes.addedCombatantFull.exec(log);
+      const matches = m?.groups;
```

Comment:
Should this be `if (!(matches && matches.groups))` similar to timeline or should timeline adopt this style?

Reply:
I think part of it is that trigger functions in all trigger files and the `AddMonster` function here take `matches` (but is really `groups`).  And so, there's some code that says `matches` is what gets returned from a regex match, and there's other code that say `matches` is the `groups` parameter.  I was trying to make this just match the parameter here.

Do you think everything should adopt this style? (I'm not sure what to call `m` though.)

Reply:
I don't have strong opinions here; I saw it as a means for one pattern to absorb the other and be an emergent pattern going forward for anyone that might run into a similar pattern of code. For Timeline, I think the names were already somewhat useful? Or, if not, I tried to make them useful:
```
commandMatch = regexes.syncCommand.exec(line);
```
And I would expect that from there if you wanted matches to be that, or you wanted that to be groups, you could reassign from there.

Additionally, we can ignore this comment and see what shakes out in further conversions.

Reply:
Yeah, I think this is a larger conversation to have to clean up bits in multiple places.